### ISSUE TITLE
Streaming extension

### ISSUE THREAD

**Issue by tdeenes** [Nicoshev/rapidhash#3]
@Nicoshev Do you plan to add streaming capability in a similar vein to `XXH64_update`? My use case would be to calculate the hash of a file incrementally, e.g., like this [hashFile](https://github.com/Cyan4973/xxHash/blob/a57f6cce2698049863af8c25787084ae0489d849/xxhash.h#L147) example.

**Comment by trapexit** [comment]
If rapidhash is like xxhash it really can't do streaming. The value, at least of xxhash, will change depending on block size. But of course that is for @Nicoshev to comment on authoritatively.

**Comment by Nicoshev** [comment]
I think it is possible to achieve. The complexity to do so derives from whether we require the final hash to be equal as if the whole buffer had been hashed at once.

If hashes can be different, then we can just call the hash function repeatedly. We would need to test ways to maintain a main seed that does not cause the hash quality to diverge as calls go by.

If the final hash needs to be equal to hashing the buffer at once, we need to maintain a state containing: the total bytes hashed so far, the number of outstanding bytes without hashing and 3 seeds.
With that information we can hash chunks of 48 bytes independentlly and then have a separate rapidhash_finalize function that computes the last bytes.
This finalize function will either call regular rapidhash if up to 16 bytes have been requested to process and the piece of code that comes after the loop otherwise.
The most challenging part of this approach is to find a way to reuse code while guaranteeing no performance penalty is introduced to the main hash function.

Maybe the best way is to encapsulate the compact loop code in a function and call it from the main rapidhash function.
The function should get inlined; but if not, utilizing the compact loop already trades-off performance, so we would be hurting the case that tends to be the less performant.
The update function would call that function if the outstanding bytes are at least 96, always leaving up to 48 bytes unprocessed.
Then, the finally function would just need to call rapidhash_internal and I believe it would work.


Cheers,
Nicolas

**Comment by tdeenes** [comment]
My understanding is that xxhash does almost the same that you described (see [here](https://github.com/Cyan4973/xxHash/blob/a57f6cce2698049863af8c25787084ae0489d849/xxhash.h#L169)), and due to the overhead, it keeps the implementation of the streaming logic separated.

**Comment by trapexit** [comment]
Sorry. I meant wyhash. Not xxhash. My bad.

**Comment by Techcable** [comment]
Due to the design of Rust's [`std::hash::Hasher` trait], this is necessary for a rust port.

**Comment by haberman** [comment]
> If hashes can be different, then we can just call the hash function repeatedly. We would need to test ways to maintain a main seed that does not cause the hash quality to diverge as calls go by.

I am interested in this use case. Will hash quality degrade if I call:

```c
  uint64_t hash = RAPID_SEED;
  hash = rapidhash_withSeed(buf1, len1, hash);
  hash = rapidhash_withSeed(buf2, len2, hash);
  hash = rapidhash_withSeed(buf3, len3, hash);
  // etc.
```

I don't need this to yield the same hash as the concatenated buffer would have, but I do want the result to have the same quality.

**Comment by amateur80lvl** [comment]
I implemented streaming version for my own needs. It's a bit simplified given that I work with specific data and don't have to worry if len < 16. But this **len** is what actually stops implementing streaming in the right way. It is involved in hash calculations but when we deal with streaming we often don't know the length. What if we omit it? I did not notice any difference in quality but I'm not an expert in this area. And if we do that, it will be just an incompatible version at a global scope which should be named accordingly. Rapidhash2, probably :)

**Comment by oertl** [comment]
@amateur80lvl this can be solved by using a small buffer. For example, the [streaming Wyhash implementations in hash4j](https://github.com/dynatrace-oss/hash4j/blob/83f97b36f4bed60c6b9edc5674fba6971c462c1f/src/main/java/com/dynatrace/hash4j/hashing/AbstractWyhashFinal.java#L285C25-L285C32) uses a fixed-size buffer of 56 bytes. I have not yet worked on it, but I expect it will work similarly for Rapidhash as it is derived from Wyhash.

**Comment by amateur80lvl** [comment]
@oertl, if you know **len** in advance, streaming API is not always necessary and can be worked around with memory mapped files, for example. As for real streaming, **len** is an obstable:

https://github.com/Nicoshev/rapidhash/blob/588978411df8683777429f729be5213eb1bfd5f3/rapidhash.h#L245

**Comment by oertl** [comment]
@amateur80lvl, you are right. I overlooked this line, which prevents a streaming implementation. The same problem existed with Wyhash (see https://github.com/wangyi-fudan/wyhash/issues/121).

**Comment by amateur80lvl** [comment]
@oertl basically, I always know size of data in my apps and can't imagine someone need a hash of indefinite data. But... scanning C strings twice is annoying.

**Comment by Nicoshev** [comment]
@amateur80lvl is right, the total hashed size in bytes is used to initialize the seed; and this may not be known in beforehand when doing stream hashing.

Removing that '^len' causes the hash to not pass SMHasher3.
When that xor is not present, appending or prepending 0s to the hashed buffer does not cause every bit to change with near 50% probability.
This results in a notably higher collision rate when prepending or appending 0s.

Appending 0s is actually more likely to happen in a streaming scenario, where input is added at the end of already hashed data.
So we need to address this property correctly, I'll think on how can we do it without knowing the final length in beforehand.

**Comment by Nicoshev** [comment]
@haberman Calling rapidhash multiple times using the previous result as seed should be mostly fine. Quality may drop a bit, but from the testing I have done, the collision probability should remain below 1/(2^63).

I'll try to do further testing to confirm this theory.

**Comment by tdeenes** [comment]
@Nicoshev  I think many of us would already benefit from a rapidhash version which would provide an interface for known `len` (or a `rapidhash_file` function directly). So if a real streaming solution turns out to be more challenging, even this pseudo-streaming extension would be helpful.

**Comment by trapexit** [comment]
Can people step back and articulate the reason for wanting a streamable version? wyhash and rapidhash are good for short data sets. For using to hash a string in a data structure for example. If the algo doesn't work for longer data sets then it doesn't work. I don't see a value in trying to shoehorn it into a usecase. Decisions made to be fast on short arrays of data may be incompatible with something intended for more traditional uses.

**Comment by oertl** [comment]
If a hash algorithm is not streamable and you need to hash more complicated objects, they must be serialized to a byte buffer first. This may require the allocation of a byte buffer or using a static one. Even worse, sometimes you don't know the size in advance, such that the byte buffer must grow during serialization or the total length must be determined in a previous iteration. With a streamable algorithm, the fields and nested objects can be directly fed into the hash algorithm without additional memory allocations, copying data, or multiple roundtrips. Consider hashing an object with multiple null-terminated string fields as a simple example.

**Comment by oertl** [comment]
> It isn't designed for that.

Rapidhash is the official successor to Wyhash, which can be implemented to work with streams. Therefore, it is not unreasonable to expect the same from Rapidhash.

**Comment by trapexit** [comment]
> which can be implemented to work with streams. 

* https://github.com/wangyi-fudan/wyhash/blob/master/wyhash.h

I see no wyhash_update(). It was added for one or two releases and then removed.

* https://github.com/wangyi-fudan/wyhash/issues/131

**Comment by Nicoshev** [comment]
If the '^len' operation is removed from the seed initialization routine, and processing of variable 'b' on line 292 gets changed to 'b^=seed^len;', the function becomes streamable and still passes both SMHasher and SMHasher3.

It does become a little bit worse regarding short input speed nevertheless, and I cannot confirm yet that it is also of higher quality than wyhash.

I do not know if it's worth to take a hit in speed and possibly quality just for making the function streamable.

**Comment by thynson** [comment]
> Due to the design of Rust's [std::hash::Hasher trait](https://doc.rust-lang.org/std/hash/trait.Hasher.html), this is necessary for a rust port.

Nope, according to the doc:

> Nor can you assume that adjacent write calls are merged, so it’s possible, for example, that
>
>    ```rust
>     hasher.write(&[1, 2]);
>     hasher.write(&[3, 4, 5, 6]);
>     and
>     
>     hasher.write(&[1, 2, 3, 4]);
>     hasher.write(&[5, 6]);
>    ```
>end up producing different hashes.

So it's valid to implement a `std::hash::Hasher` via `self.seed  = rapidhash_withSeed(slice, self.seed);`;
and from security perspective I personally feel such implementation resists from crafted tuples that lead to collisions.

**Comment by thynson** [comment]
> If the '^len' operation is removed from the seed initialization routine, and processing of variable 'b' on line 292 gets changed to 'b^=seed^len;', the function becomes streamable and still passes both SMHasher and SMHasher3.
> 
> It does become a little bit worse regarding short input speed nevertheless, and I cannot confirm yet that it is also of higher quality than wyhash.
> 
> I do not know if it's worth to take a hit in speed and possibly quality just for making the function streamable.

I think it's worth a change as long as it pass SMHasher and SMHasher3.

Alternatively we could provide a streaming variant. And for hash tables, the construction of `self.seed = rapidhash_withSeed(slice, self.seed);` should work well in performance (or even better,  due to buffering can be totally eliminated), although I'm not sure if there is any quality degration.

**Comment by trapexit** [comment]
1) If the performance of the hash (in any dimension) is worse then there should be no replacement of that hash algo. The original intent and usage is for small strings/data and that is what it should stay focused on IMO.
2) Unless a streaming variant is faster or has other better attributes when compared to other hashes... which can't be established yet since no algo has been proposed... there is no point to a streaming variant. What's the point of having an algo if it doesn't excel at something?

**Comment by tansy** [comment]
Because the hash is
> using len to initialize the seed

there is no way to make it streaming, as one doesn't know the length of data beforehand, despite [claims](https://github.com/Nicoshev/rapidhash/issues/3#issuecomment-2298199746) that you do.

**Comment by hoxxep** [comment]
Small update, I've built a streaming version of rapidhash in the rust crate. It still needs to know the length up front, and so this implementation takes a rust `&mut File` type as input and queries the file metadata, but in theory it could be adjusted to work over any stream that's compatible with the `std::io::Read` trait and has a known length. The rust crate's CLI can hash many GB files with 1.0MB of memory usage (will be released in v1.3.0).

It makes use of the rust slice type and reads 96 byte chunks into an internal 192 byte buffer. Future work could replace the use of `BufReader` to manage a larger buffer internally and avoid the double-copying (read -> bufreader -> internal buffer).

**Implementation:** https://github.com/hoxxep/rapidhash/blob/954c6bd63a72f6f6f1c035734499a6584deaa446/src/rapid_file.rs
